src/browser/NativeMessagingHost.cpp

```diff
@@ -0,0 +1,298 @@
+/*
+*  Copyright (C) 2017 KeePassXC Team <team@keepassxc.org>
+*
+*  This program is free software: you can redistribute it and/or modify
+*  it under the terms of the GNU General Public License as published by
+*  the Free Software Foundation, either version 3 of the License, or
+*  (at your option) any later version.
+*
+*  This program is distributed in the hope that it will be useful,
+*  but WITHOUT ANY WARRANTY; without even the implied warranty of
+*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+*  GNU General Public License for more details.
+*
+*  You should have received a copy of the GNU General Public License
+*  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <QMutexLocker>
+#include <QtNetwork>
+#include <iostream>
+#include "sodium.h"
+#include "NativeMessagingHost.h"
+#include "BrowserSettings.h"
+
+#ifndef Q_OS_LINUX
+#if defined Q_OS_MAC || defined Q_OS_UNIX
+#include <sys/types.h>
+#include <sys/event.h>
+#include <sys/time.h>
+#include <unistd.h>
+#endif
+#endif
+
+#ifdef Q_OS_LINUX
+#include <sys/epoll.h>
+#include <unistd.h>
+#endif
+
+#ifdef Q_OS_WIN
+#include <fcntl.h>
+#include <io.h>
+#endif
+
+NativeMessagingHost::NativeMessagingHost(DatabaseTabWidget* parent) :
+    m_mutex(QMutex::Recursive),
+    m_browserClients(m_browserService),
+    m_browserService(parent)
+{
+#ifndef Q_OS_WIN
+    m_notifier.reset(new QSocketNotifier(fileno(stdin), QSocketNotifier::Read, this));
+    connect(m_notifier.data(), SIGNAL(activated(int)), this, SLOT(newNativeMessage()));
+#endif
+    m_localServer.reset(new QLocalServer(this));
+    m_localServer->setSocketOptions(QLocalServer::UserAccessOption);
+#ifdef Q_OS_WIN
+    m_running.store(false);
+    _setmode(_fileno(stdin), _O_BINARY);
+    _setmode(_fileno(stdout), _O_BINARY);
+#endif
+    if (BrowserSettings::isEnabled() && !m_running) {
+        run();
+    }
+
+    connect(&m_browserService, SIGNAL(databaseLocked()), this, SLOT(databaseLocked()));
+    connect(&m_browserService, SIGNAL(databaseUnlocked()), this, SLOT(databaseUnlocked()));
+}
+
+NativeMessagingHost::~NativeMessagingHost()
+{
+    stop();
+}
+
+int NativeMessagingHost::init()
+{
+    QMutexLocker locker(&m_mutex);
+    return sodium_init();
+}
+
+void NativeMessagingHost::run()
+{
+    QMutexLocker locker(&m_mutex);
+    if (!m_running.load() && init() == -1) {
+        return;
+    }
+
+    // Update KeePassXC/keepassxc-proxy binary paths to Native Messaging scripts
+    if (BrowserSettings::updateBinaryPath()) {
+        BrowserSettings::updateBinaryPaths(BrowserSettings::useCustomProxy() ? BrowserSettings::customProxyLocation() : "");
+    }
+
+    m_running.store(true);
+#ifdef Q_OS_WIN
+    m_future = QtConcurrent::run(this, &NativeMessagingHost::readNativeMessages);
+#endif
+
+    if (BrowserSettings::supportBrowserProxy()) {
+#ifndef Q_OS_WIN
+        QString serverPath = "/tmp/kpxc_server";
```

Comment:
How about ```$XDG_RUNTIME_DIR/kpxc_server``` instead of /tmp/kpxc_server? /tmp is really cluttered on systems with many applications.

Reply:
Good point.

Reply:
I agree. This would be a much better choice.